File: backend/app/schemas/task.py
"""
任务数据验证模型
"""
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from app.models.task import TaskStatus


class ImageFileResponse(BaseModel):
    """图片文件响应"""
    id: int = Field(..., description="图片ID")
    original_filename: str = Field(..., description="原始文件名")
    file_path: str = Field(..., description="文件路径")
    file_size: int = Field(..., description="文件大小")
    file_type: str = Field(..., description="文件类型")
    created_at: datetime = Field(..., description="创建时间")

    model_config = {"from_attributes": True}


class AnalysisResultResponse(BaseModel):
    """分析结果响应"""
    id: int = Field(..., description="分析结果ID")
    agent_type: str = Field(..., description="Agent类型")
    page_type: Optional[str] = Field(None, description="页面类型")
    elements_json: Optional[str] = Field(None, description="UI元素JSON")
    interactions_json: Optional[str] = Field(None, description="交互JSON")
    analysis_summary: Optional[str] = Field(None, description="分析总结")
    created_at: datetime = Field(..., description="创建时间")

    model_config = {"from_attributes": True}


class ScriptResponse(BaseModel):
    """脚本响应"""
    id: int = Field(..., description="脚本ID")
    script_type: str = Field(..., description="脚本类型")
    script_content: str = Field(..., description="脚本内容")
    script_language: str = Field(..., description="脚本语言")
    created_at: datetime = Field(..., description="创建时间")

    model_config = {"from_attributes": True}


class UIElementResponse(BaseModel):
    """统一UI元素响应"""
    id: int = Field(..., description="元素ID")
    name: str = Field(..., description="元素名称")
    type: str = Field(..., description="元素类型")
    text: Optional[str] = Field(None, description="元素文本")
    source: str = Field(..., description="来源: vision/dom/merge")
    locator: Optional[str] = Field(None, description="最佳定位器")
    xpath: Optional[str] = Field(None, description="XPath")
    css_selector: Optional[str] = Field(None, description="CSS选择器")
    element_id: Optional[str] = Field(None, description="元素id属性")
    element_class: Optional[str] = Field(None, description="元素class属性")
    element_name: Optional[str] = Field(None, description="元素name属性")
    placeholder: Optional[str] = Field(None, description="placeholder")
    href: Optional[str] = Field(None, description="href")
    aria_label: Optional[str] = Field(None, description="aria-label")
    role: Optional[str] = Field(None, description="role")
    data_testid: Optional[str] = Field(None, description="data-testid")
    page_url: Optional[str] = Field(None, description="页面URL")
    confidence: Optional[float] = Field(None, description="置信度")

    model_config = {"from_attributes": True}


class PageElementResponse(BaseModel):
    """页面元素响应（Playwright抓取的DOM元素，兼容旧接口）"""
    id: int = Field(..., description="元素ID")
    tag_name: str = Field(..., description="标签名")
    element_text: Optional[str] = Field(None, description="元素文本")
    element_id: Optional[str] = Field(None, description="元素ID属性")
    element_class: Optional[str] = Field(None, description="元素class属性")
    element_name: Optional[str] = Field(None, description="元素name属性")
    placeholder: Optional[str] = Field(None, description="placeholder属性")
    href: Optional[str] = Field(None, description="href属性")
    aria_label: Optional[str] = Field(None, description="aria-label属性")
    role: Optional[str] = Field(None, description="role属性")
    xpath: Optional[str] = Field(None, description="XPath定位")
    css_selector: Optional[str] = Field(None, description="CSS选择器定位")

    model_config = {"from_attributes": True}


class TaskCreate(BaseModel):
    """创建任务请求"""
    task_name: str = Field(..., min_length=1, max_length=255, description="任务名称")


class TaskResponse(BaseModel):
    """任务响应"""
    id: int = Field(..., description="任务ID")
    task_name: str = Field(..., description="任务名称")
    status: TaskStatus = Field(..., description="任务状态")
    input_mode: str = Field(default="image", description="输入模式: image/url")
    task_type: Optional[str] = Field(default="web", description="测试类型: web/api/performance/android")
    type_config: Optional[str] = Field(None, description="测试类型配置(JSON)")
    page_url: Optional[str] = Field(None, description="页面URL")
    error_message: Optional[str] = Field(None, description="错误信息")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    # 关联数据
    images: List[ImageFileResponse] = Field(default_factory=list, description="图片列表")
    analysis_result: Optional[AnalysisResultResponse] = Field(None, description="分析结果")
    script: Optional[ScriptResponse] = Field(None, description="测试脚本")
    ui_elements: List[UIElementResponse] = Field(default_factory=list, description="统一元素列表")
    page_elements: List[PageElementResponse] = Field(default_factory=list, description="页面元素列表(DOM抓取,兼容)")

    model_config = {"from_attributes": True}


class TaskListResponse(BaseModel):
    """任务列表响应"""
    total: int = Field(..., description="总数")
    items: list[TaskResponse] = Field(..., description="任务列表")

File: frontend/src/services/task.ts
import request from './request';

export interface UIElement {
  id: number;
  name: string;
  type: string;
  text: string | null;
  source: 'vision' | 'dom' | 'merge';
  locator: string | null;
  xpath: string | null;
  css_selector: string | null;
  element_id: string | null;
  element_class: string | null;
  element_name: string | null;
  placeholder: string | null;
  href: string | null;
  aria_label: string | null;
  role: string | null;
  data_testid: string | null;
  page_url: string | null;
  confidence: number | null;
}

export interface Task {
  id: number;
  task_name: string;
  status: 'pending' | 'processing' | 'success' | 'failed';
  input_mode: 'image' | 'url';
  page_url: string | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
  images: Array<{
    id: number;
    original_filename: string;
    file_path: string;
    file_size: number;
    file_type: string;
  }>;
  analysis_result: {
    id: number;
    agent_type: string;
    page_type: string | null;
    elements_json: string | null;
    interactions_json: string | null;
    analysis_summary: string | null;
  } | null;
  script: {
    id: number;
    script_type: string;
    script_content: string;
    script_language: string;
  } | null;
  ui_elements: UIElement[];
  page_elements: Array<{
    id: number;
    tag_name: string;
    element_text: string | null;
    element_id: string | null;
    element_class: string | null;
    element_name: string | null;
    placeholder: string | null;
    href: string | null;
    aria_label: string | null;
    role: string | null;
    xpath: string | null;
    css_selector: string | null;
  }>;
}

export interface TaskListResponse {
  total: number;
  items: Task[];
}

// 上传图片并创建任务（统一入口，使用原生fetch避免axios的Content-Type问题）
export async function uploadAndCreateTask(formData: FormData) {
  const response = await fetch('/api/tasks/upload', {
    method: 'POST',
    body: formData,
    credentials: 'include',
  });
  if (!response.ok) throw new Error(`上传失败 (${response.status})`);
  const res = await response.json();
  return res;
}

// URL模式创建任务（统一入口）
export function createTaskByUrl(url: string, task_name?: string) {
  return request.post('/tasks/create', { url, task_name });
}

// 获取任务列表
export function getTaskList(skip = 0, limit = 20) {
  return request.get('/tasks/', { params: { skip, limit } });
}

// 获取任务详情
export function getTask(id: number) {
  return request.get(`/tasks/${id}`);
}

// 删除任务
export function deleteTask(id: number) {
  return request.delete(`/tasks/${id}`);
}

// 重新执行任务
export function rerunTask(id: number) {
  return request.post(`/tasks/${id}/rerun`);
}

// 下载脚本
export function downloadScript(id: number) {
  window.open(`/api/tasks/${id}/download`, '_blank');
}

// 更新脚本内容
export function updateScript(taskId: number, scriptContent: string) {
  return request.put(`/tasks/${taskId}/script`, { script_content: scriptContent });
}

// ============ 执行管理 ============

export interface ExecutionRecord {
  id: number;
  task_id: number;
  status: 'pending' | 'running' | 'success' | 'failed';
  start_time: string | null;
  end_time: string | null;
  duration: number | null;
  success_count: number;
  failed_count: number;
  error_message: string | null;
  log_content: string | null;
  report_path: string | null;
  screenshot_path: string | null;
  created_at: string;
}

// 执行脚本
export function executeScript(taskId: number) {
  return request.post(`/executions/${taskId}/execute`);
}

// 获取执行记录
export function getExecution(executionId: number) {
  return request.get(`/executions/${executionId}`);
}

// 获取任务的执行记录列表
export function getTaskExecutions(taskId: number) {
  return request.get(`/executions/task/${taskId}/list`);
}

// 获取执行日志
export function getExecutionLogs(executionId: number) {
  return request.get(`/executions/${executionId}/logs`, { responseType: 'text' });
}

// 获取报告URL
export function getReportUrl(executionId: number) {
  return `/api/executions/${executionId}/report`;
}

// 获取截图URL
export function getScreenshotUrl(executionId: number) {
  return `/api/executions/${executionId}/screenshot`;
}
